# 4.8 Grouping Data & Aggregating Variables

### This script contains the following points:
1. Importing Libraries
2. Importing Data 
3. Checking Data
4. Task
5. Exporting Data

# 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# Project folder path
path=r'C:\Users\maryn\Documents\Data Projects\Instacart Basket Analysis'

In [3]:
# Import dataset orders_products_merged_derived.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

# 03. Checking Data

In [4]:
# Display the information orders_products_merged.pkl
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days        Average orders  
1  Regularly busy         Slowest days        Average orders  
2  Regularly busy         Slowest days           Most orders  
3      Least busy         Slowest days        Average orders  
4      Least busy         Slowest days           Most orders

In [5]:
ords_prods_merge.shape

(32404859, 18)

# 04. Task

## Step 2 Find the aggregated mean of the “order_number” column grouped by “department_id” for the entire dataframe. 

In [6]:
# Performing a Single Aggregation to produce a single descriptive statistic for the “order_number” column
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

#### The results of the aggregated mean for a subset of 1000000 entries and for the whole dataframe of 32404859 entries are different. The data in the subset is not a fully representative sample of the dataframe.

## Step 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [13]:
# Split the data into groups based on the “user_id” column
# Apply the transform() function on the “order_number” column to generate the maximum orders for each user
# Create a new column, “max_order,” into which you’ll place the results of your aggregation.
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [14]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regularly busy days        Average orders         10  
1  Regularly busy         Slowest days        Average orders         10  
2  Regularly busy         Slowest days           Most orders         10  
3      Least busy         Slowest days        Average orders         10  
4      Least busy         Slowest days           Most orders         10

In [15]:
# Create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value.
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# Print frequency of the new "loyalty_flag" column using the value_counts() function 
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [19]:
# Check the outcome
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

## Step 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. 

### Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers..”

In [20]:
# Check basic, multiple statistics about 3 types of customers and their spending habits 
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### Loyal customers spend less on average than the other two types of customer, while the average check for new customers is the most expensive.

# Detection of outliers (Part 4.9)

In [32]:
# Exploratory checks for detection of outliers
ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
10030345                 14                     5.0       21553   
10030346                 16                     4.0       21553   
10030347                 21                    13.0       21553   
10030348                  8                    12.0       21553   
10030349                 19                    30.0       21553   
...                     ...                     ...         ...   
29166209                  8                     4.0       33664   
29166210                  9                     2.0       33664   
29166211                 15                     8.0       33664   
29166212                 20                     4.0       33664   
29166213                 11                     4.0       33664   

          add_to_cart_order  reordered                      product_name  ...  \
10030345                  5          0  Lowfat 2% Milkfat Cottage Cheese  ...   
10030346                  3          1  Lowfat 2% Milkfat Cottage Cheese  ...   
10030347                  6          0  Lowfat 2% Milkfat Cottage Cheese  ...   
10030348                  3          1  Lowfat 2% Milkfat Cottage Cheese  ...   
10030349                  9          0  Lowfat 2% Milkfat Cottage Cheese  ...   
...                     ...        ...                               ...  ...   
29166209                  1          0             2 % Reduced Fat  Milk  ...   
29166210                  1          1             2 % Reduced Fat  Milk  ...   
29166211                 25          0             2 % Reduced Fat  Milk  ...   
29166212                  8          0             2 % Reduced Fat  Milk  ...   
29166213                 23          1             2 % Reduced Fat  Milk  ...   

             price_range_loc     busiest_day         busiest_days  \
10030345  High-range product  Regularly busy  Regularly busy days   
10030346  High-range product  Regularly busy  Regularly busy days   
10030347  High-range product  Regularly busy  Regularly busy days   
10030348  High-range product     Busiest day         Busiest days   
10030349  High-range product  Regularly busy         Busiest days   
...                      ...             ...                  ...   
29166209  High-range product     Busiest day         Busiest days   
29166210  High-range product     Busiest day         Busiest days   
29166211  High-range product  Regularly busy         Slowest days   
29166212  High-range product  Regularly busy         Slowest days   
29166213  High-range product     Busiest day         Busiest days   

         busiest_period_of_day max_order      loyalty_flag     spending  \
10030345           Most orders        40  Regular customer   108.648299   
10030346           Most orders        40  Regular customer   108.648299   
10030347        Average orders         4      New customer  1154.792308   
10030348        Average orders         4      New customer  1154.792308   
10030349        Average orders        16  Regular customer   114.426619   
...                        ...       ...               ...          ...   
29166209        Average orders        39  Regular customer  1106.743956   
29166210           Mos

### The prices 14,900.00 and $99,999.00 are the values that skew the data. There are 5,127 rows with outlier observations (prices greater than 100).s.

In [33]:
# Mark outliers as missing because they don't make sense with the other values in the column.
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [34]:
# Check the current max value 
ords_prods_merge['prices'].max()

25.0

## Step 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. .”

### Create a spending flag for each user based on the average price across all their orders using the following criteria:
1) If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
2) If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [21]:
# Create a new colums 'spending' aggregated the mean of 'prices' value for each user ID.
ords_prods_merge['spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [22]:
# Check the outcome
ords_prods_merge[['user_id','prices','spending']].head(20)

user_id  prices  spending
0         1     9.0  6.367797
1         1     9.0  6.367797
2         1     9.0  6.367797
3         1     9.0  6.367797
4         1     9.0  6.367797
5         1     9.0  6.367797
6         1     9.0  6.367797
7         1     9.0  6.367797
8         1     9.0  6.367797
9         1     9.0  6.367797
10       15     9.0  3.980556
11       15     9.0  3.980556
12       15     9.0  3.980556
13       15     9.0  3.980556
14       15     9.0  3.980556
15       19     9.0  8.372059
16       19     9.0  8.372059
17       19     9.0  8.372059
18       21     9.0  7.012683
19       31     9.0  8.137793

In [23]:
# Create a flag that assigns a “Spending_flag” label to a user ID based on its corresponding mean of 'prices' value.
ords_prods_merge.loc[ords_prods_merge['spending'] < 10, 'spending_flag'] = 'Low spender'

In [24]:
ords_prods_merge.loc[ords_prods_merge['spending'] >= 10, 'spending_flag'] = 'High spender'

In [25]:
# Print frequency of the new "spending_flag" column using the value_counts() function 
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

### Among customers, the group of low spenders is 50 times larger than the group of high spenders.

## Step 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

### The criteria for the flag should be as follows:
1) If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a"“Non-frequent custome"
2)  If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a"“Regular custome"
3)  If the median is lower than or equal to 10, then the customer should be labeled a"“Frequent custome"”

In [26]:
# Create a new colums 'ordering_behavior' aggregated the median of 'days_since_prior_order' value for each user ID.
ords_prods_merge['ordering_behavior'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [27]:
# Check the outcome
ords_prods_merge[['user_id','days_since_prior_order','ordering_behavior']].head(20)

user_id  days_since_prior_order  ordering_behavior
0         1                     NaN               20.5
1         1                    15.0               20.5
2         1                    21.0               20.5
3         1                    29.0               20.5
4         1                    28.0               20.5
5         1                    19.0               20.5
6         1                    20.0               20.5
7         1                    14.0               20.5
8         1                     0.0               20.5
9         1                    30.0               20.5
10       15                     7.0               10.0
11       15                     8.0               10.0
12       15                     7.0               10.0
13       15                     7.0               10.0
14       15                    14.0               10.0
15       19                     6.0                6.0
16       19                     8.0                6.0
17       19                     6.0                6.0
18       21                     6.0                7.5
19       31                     7.0                6.0

In [28]:
# Create a flag that assigns a "order frequency" label to a user ID based on its corresponding median order frequency value 
ords_prods_merge.loc[ords_prods_merge['ordering_behavior'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['ordering_behavior'] > 10) & (ords_prods_merge['ordering_behavior'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['ordering_behavior'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [31]:
# Print frequency of the new "order_freq_flag" column using the value_counts() function 
ords_prods_merge['order_freq_flag'].value_counts(dropna = False)

order_freq_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
nan                             5
Name: count, dtype: int64

### More than half of the customers are frequent customers. There are 5 missing values, which is assumed because there is not enough data to classify 5 users in the system. This may be due to missing values in the column 'days_since_prior_order' and if the customer has only ordered once.

# 05. Exporting Data

In [35]:
# Exporting ords_prods_merge as a orders_products_merged_agg.pkl in pickle format
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_agg.pkl'))